### PR TITLE
caasoperator: add SetContainerSpec

### PR DESCRIPTION
**Pull request by axw**
## Description of change

Add methods to state.CAASModel for getting,
setting, and watching the container spec for
applications and units in CAAS models.

Fix a bug when creating applications in
CAAS models, causing the series to not
be inherited from the charm.

Add basic support for adding units in a
CAAS model. Very much incomplete,
just enough for testing the basics.

Add a SetContainerSpec method to the
CAASOperator API facade and client.
Add an ContainerSpecSetter interface
as a dependency of the caasoperator
worker (currently unused).

## QA steps

smoke test IAAS model  / deploy, to make sure nothing is broken

## Documentation changes

None.

## Bug reference

None.

### CODE REVIEW

File: api/caasoperator/client_test.go [hunk 1]
```diff
@@ -166,3 +167,40 @@ func (s *operatorSuite) TestWatchApplicationConfig(c *gc.C) {
...
+	})
+
+	client := caasoperator.NewClient(apiCaller)
+	err := client.SetContainerSpec(tag.Id(), "spec")
+	c.Assert(err, gc.ErrorMatches, "bletch")
+}
+
+func (s *operatorSuite) TestSetContainerSpecInvalidEntityame(c *gc.C) {
+	client := caasoperator.NewClient(basetesting.APICallerFunc(func(_ string, _ int, _, _ string, _, _ interface{}) error {
+		return errors.New("should not be called")
```
Comment (wallyworld) [inline api/caasoperator/client_test.go]: c.Fail()?
Guess it doesn't matter

File: state/watcher.go [hunk 2]
```diff
@@ -3148,3 +3148,33 @@ func (w *externalControllersWatcher) loop() error {
 		}
 	}
 }
+
+// WatchContainerSpec returns a watcher observing changes that affect the meter status
```
Comment (wallyworld) [inline state/watcher.go]: typo

File: state/state.go [hunk 3]
```diff
@@ -1332,6 +1334,24 @@ func (st *State) processIAASModelApplicationArgs(args *AddApplicationArgs) error
...
+	if args.Series == "" {
+		// args.Series is not set, so use the series in the URL.
+		args.Series = args.Charm.URL().Series
+		if args.Series == "" {
+			// Should not happen, but just in case.
+			return errors.New("series is empty")
+		}
+	}
+
+	// TODO(caas) restrict the series to CAAS series.
```
Comment (wallyworld) [inline state/state.go]: could we make all of the TODOs "TODO(caas)" so we can easily search

File: state/watcher.go (hunk 2 above)
Comment (axw) [reply state/watcher.go]: fixed

File: state/state.go (hunk 3 above)
Comment (axw) [reply state/state.go]: done

File: api/caasoperator/client_test.go (hunk 1 above)
Comment (axw) [reply api/caasoperator/client_test.go]: doesn't matter - it'll be clear why it fails, if it ever does

**Comment by axw** [comment]
$$merge$$

**Pull request merged**